src/js/modules/core/services/bitcoin-link/bitcoin-link.service.js

```diff
@@ -0,0 +1,96 @@
+
+            var protocol = qregex[1];
+            var address = qregex[2];
+            var query = qregex[4];
+
+            var options = parseQuery("?"+query);
+            if (options.amount) {
+                options.amount = Number(options.amount);
+                if (!isFinite(options.amount)) throw new Error('Invalid amount');
+                if (options.amount < 0) throw new Error('Invalid amount');
+            }
+
+            return { address: address, options: options, protocol: protocol};
+        }
+
+        function parse(bitcoinLink) {
+            var deferred = $q.defer();
+
+            try {
+                var uri = decodeBitcoinLink(bitcoinLink);
+            } catch (e) {
+                deferred.reject('Unable to decode bitcoin link. May be corrupted');
+                return deferred.promise;
+            }
+
+            var res = {};
+            res.network = uri.protocol;
+
+            // BIP70
+            if (uri && uri.options && uri.options.r) {
+                var paymentUrl = uri.options.r;
+                // . ..
+                var validation = new bip70.X509.RequestValidator({
+                    trustStore: bip70.X509.TrustStore
+                });
+
+                var client = new bip70.HttpClient();
+
+                var network;
+                var networkConfig;
+                if (uri.protocol == 'bitcoin') {
+                    networkConfig = bip70.NetworkConfig.Bitcoin();
+                    network = bitcoinJS.networks.bitcoin;
+                    // TODO: BCH BIP70 is currently incompatible with BitPay
+                } else {
+                    $timeout(function() {
+                        $cordovaToast.showLongCenter($translate.instant('MSG_INVALID_RECIPIENT'.sentenceCase()));
+                    }, 350);
+                    deferred.reject('Unsupported network for BIP70 requests');
+                }
+
+                client.getRequest(paymentUrl, validation, networkConfig)
+                    .then(function(request) {
+                        var details = bip70.ProtoBuf.PaymentDetails.decode(request[0].serializedPaymentDetails);
+                        if (details.outputs.length > 1) {
+                            deferred.reject("Multiple output payment requests are not supported");
+                        }
+                        res.recipientAddress = bitcoinJS.address.fromOutputScript(blocktrailSDK.Buffer.from(details.outputs[0].script), network);
+                        res.recipientSource = 'BIP70PaymentURL';
+                        res.recipientDisplay = details.memo;
+                        res.btcValue = parseFloat(blocktrailSDK.toBTC(details.outputs[0].amount));
+                        deferred.resolve(res);
+                    }, function(err) {
+                        console.log("err - abort request");
+                        console.log(err.message);
+                        $state.go('app.wallet.summary');
```

Comment:
$state in here might be a bit much, and the caller never gets told about the issue. maybe reject the promise, and move the $state.go() back to when the rejection is handled by the caller?